File: 03-searching-and-sorting/01_binary_search/binary_search.py
"""
Problem: Binary Search
LeetCode: #704 (Easy)

Time Complexity: O(log N)
Space Complexity: O(1)
"""

from typing import List

def binary_search(nums: List[int], target: int) -> int:
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

if __name__ == "__main__":
    print("[Python] Binary Search Test")
    tests = [
        ([-1, 0, 3, 5, 9, 12], 9, 4),
        ([-1, 0, 3, 5, 9, 12], 2, -1),
    ]
    for nums, target, expected in tests:
        res = binary_search(nums, target)
        print(f"nums: {nums}, target: {target} -> Index: {res} | Expected: {expected}")
        assert res == expected
    print("All Binary Search tests passed!")

File: 03-searching-and-sorting/01_binary_search/binary_search.cpp
/**
 * Problem: Binary Search
 * LeetCode: #704 (Easy)
 * 
 * Time Complexity: O(log N)
 * Space Complexity: O(1)
 */

#include <iostream>
#include <vector>

int search(const std::vector<int>& nums, int target) {
    int left = 0;
    int right = static_cast<int>(nums.size()) - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (nums[mid] == target) return mid;
        else if (nums[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

int main() {
    std::cout << "[C++] Binary Search Test\n";
    std::vector<int> nums = {-1, 0, 3, 5, 9, 12};
    int target = 9;
    std::cout << "Target " << target << " found at index: " << search(nums, target) << "\n";
    return 0;
}
